// @eth-optimism/contracts-bedrock/dist/types/contracts/echidna/FuzzBurn.sol/EchidnaFuzzBurnEth.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../../common";

export interface EchidnaFuzzBurnEthInterface extends utils.Interface {
  functions: {
    "echidna_burn_eth()": FunctionFragment;
    "testBurn(uint256)": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic: "echidna_burn_eth" | "testBurn"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "echidna_burn_eth",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "testBurn",
    values: [PromiseOrValue<BigNumberish>]
  ): string;

  decodeFunctionResult(
    functionFragment: "echidna_burn_eth",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "testBurn", data: BytesLike): Result;

  events: {};
}

export interface EchidnaFuzzBurnEth extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: EchidnaFuzzBurnEthInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    echidna_burn_eth(overrides?: CallOverrides): Promise<[boolean]>;

    testBurn(
      _value: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;
  };

  echidna_burn_eth(overrides?: CallOverrides): Promise<boolean>;

  testBurn(
    _value: PromiseOrValue<BigNumberish>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    echidna_burn_eth(overrides?: CallOverrides): Promise<boolean>;

    testBurn(
      _value: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {};

  estimateGas: {
    echidna_burn_eth(overrides?: CallOverrides): Promise<BigNumber>;

    testBurn(
      _value: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    echidna_burn_eth(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    testBurn(
      _value: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;
  };
}
