// @eth-optimism/contracts-bedrock/dist/types/contracts/L1/OptimismPortal.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PayableOverrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type {
  FunctionFragment,
  Result,
  EventFragment,
} from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../common";

export declare namespace Types {
  export type WithdrawalTransactionStruct = {
    nonce: PromiseOrValue<BigNumberish>;
    sender: PromiseOrValue<string>;
    target: PromiseOrValue<string>;
    value: PromiseOrValue<BigNumberish>;
    gasLimit: PromiseOrValue<BigNumberish>;
    data: PromiseOrValue<BytesLike>;
  };

  export type WithdrawalTransactionStructOutput = [
    BigNumber,
    string,
    string,
    BigNumber,
    BigNumber,
    string
  ] & {
    nonce: BigNumber;
    sender: string;
    target: string;
    value: BigNumber;
    gasLimit: BigNumber;
    data: string;
  };

  export type OutputRootProofStruct = {
    version: PromiseOrValue<BytesLike>;
    stateRoot: PromiseOrValue<BytesLike>;
    messagePasserStorageRoot: PromiseOrValue<BytesLike>;
    latestBlockhash: PromiseOrValue<BytesLike>;
  };

  export type OutputRootProofStructOutput = [string, string, string, string] & {
    version: string;
    stateRoot: string;
    messagePasserStorageRoot: string;
    latestBlockhash: string;
  };
}

export interface OptimismPortalInterface extends utils.Interface {
  functions: {
    "BASE_FEE_MAX_CHANGE_DENOMINATOR()": FunctionFragment;
    "ELASTICITY_MULTIPLIER()": FunctionFragment;
    "FINALIZATION_PERIOD_SECONDS()": FunctionFragment;
    "GUARDIAN()": FunctionFragment;
    "INITIAL_BASE_FEE()": FunctionFragment;
    "L2_ORACLE()": FunctionFragment;
    "MAXIMUM_BASE_FEE()": FunctionFragment;
    "MAX_RESOURCE_LIMIT()": FunctionFragment;
    "MINIMUM_BASE_FEE()": FunctionFragment;
    "TARGET_RESOURCE_LIMIT()": FunctionFragment;
    "depositTransaction(address,uint256,uint64,bool,bytes)": FunctionFragment;
    "donateETH()": FunctionFragment;
    "finalizeWithdrawalTransaction((uint256,address,address,uint256,uint256,bytes))": FunctionFragment;
    "finalizedWithdrawals(bytes32)": FunctionFragment;
    "initialize(bool)": FunctionFragment;
    "isOutputFinalized(uint256)": FunctionFragment;
    "l2Sender()": FunctionFragment;
    "params()": FunctionFragment;
    "pause()": FunctionFragment;
    "paused()": FunctionFragment;
    "proveWithdrawalTransaction((uint256,address,address,uint256,uint256,bytes),uint256,(bytes32,bytes32,bytes32,bytes32),bytes[])": FunctionFragment;
    "provenWithdrawals(bytes32)": FunctionFragment;
    "unpause()": FunctionFragment;
    "version()": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "BASE_FEE_MAX_CHANGE_DENOMINATOR"
      | "ELASTICITY_MULTIPLIER"
      | "FINALIZATION_PERIOD_SECONDS"
      | "GUARDIAN"
      | "INITIAL_BASE_FEE"
      | "L2_ORACLE"
      | "MAXIMUM_BASE_FEE"
      | "MAX_RESOURCE_LIMIT"
      | "MINIMUM_BASE_FEE"
      | "TARGET_RESOURCE_LIMIT"
      | "depositTransaction"
      | "donateETH"
      | "finalizeWithdrawalTransaction"
      | "finalizedWithdrawals"
      | "initialize"
      | "isOutputFinalized"
      | "l2Sender"
      | "params"
      | "pause"
      | "paused"
      | "proveWithdrawalTransaction"
      | "provenWithdrawals"
      | "unpause"
      | "version"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "BASE_FEE_MAX_CHANGE_DENOMINATOR",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "ELASTICITY_MULTIPLIER",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "FINALIZATION_PERIOD_SECONDS",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "GUARDIAN", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "INITIAL_BASE_FEE",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "L2_ORACLE", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "MAXIMUM_BASE_FEE",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "MAX_RESOURCE_LIMIT",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "MINIMUM_BASE_FEE",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "TARGET_RESOURCE_LIMIT",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "depositTransaction",
    values: [
      PromiseOrValue<string>,
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<boolean>,
      PromiseOrValue<BytesLike>
    ]
  ): string;
  encodeFunctionData(functionFragment: "donateETH", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "finalizeWithdrawalTransaction",
    values: [Types.WithdrawalTransactionStruct]
  ): string;
  encodeFunctionData(
    functionFragment: "finalizedWithdrawals",
    values: [PromiseOrValue<BytesLike>]
  ): string;
  encodeFunctionData(
    functionFragment: "initialize",
    values: [PromiseOrValue<boolean>]
  ): string;
  encodeFunctionData(
    functionFragment: "isOutputFinalized",
    values: [PromiseOrValue<BigNumberish>]
  ): string;
  encodeFunctionData(functionFragment: "l2Sender", values?: undefined): string;
  encodeFunctionData(functionFragment: "params", values?: undefined): string;
  encodeFunctionData(functionFragment: "pause", values?: undefined): string;
  encodeFunctionData(functionFragment: "paused", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "proveWithdrawalTransaction",
    values: [
      Types.WithdrawalTransactionStruct,
      PromiseOrValue<BigNumberish>,
      Types.OutputRootProofStruct,
      PromiseOrValue<BytesLike>[]
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "provenWithdrawals",
    values: [PromiseOrValue<BytesLike>]
  ): string;
  encodeFunctionData(functionFragment: "unpause", values?: undefined): string;
  encodeFunctionData(functionFragment: "version", values?: undefined): string;

  decodeFunctionResult(
    functionFragment: "BASE_FEE_MAX_CHANGE_DENOMINATOR",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "ELASTICITY_MULTIPLIER",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "FINALIZATION_PERIOD_SECONDS",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "GUARDIAN", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "INITIAL_BASE_FEE",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "L2_ORACLE", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "MAXIMUM_BASE_FEE",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "MAX_RESOURCE_LIMIT",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "MINIMUM_BASE_FEE",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "TARGET_RESOURCE_LIMIT",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "depositTransaction",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "donateETH", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "finalizeWithdrawalTransaction",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "finalizedWithdrawals",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "initialize", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "isOutputFinalized",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "l2Sender", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "params", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "pause", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "paused", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "proveWithdrawalTransaction",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "provenWithdrawals",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "unpause", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "version", data: BytesLike): Result;

  events: {
    "Initialized(uint8)": EventFragment;
    "Paused(address)": EventFragment;
    "TransactionDeposited(address,address,uint256,bytes)": EventFragment;
    "Unpaused(address)": EventFragment;
    "WithdrawalFinalized(bytes32,bool)": EventFragment;
    "WithdrawalProven(bytes32,address,address)": EventFragment;
  };

  getEvent(nameOrSignatureOrTopic: "Initialized"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "Paused"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "TransactionDeposited"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "Unpaused"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "WithdrawalFinalized"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "WithdrawalProven"): EventFragment;
}

export interface InitializedEventObject {
  version: number;
}
export type InitializedEvent = TypedEvent<[number], InitializedEventObject>;

export type InitializedEventFilter = TypedEventFilter<InitializedEvent>;

export interface PausedEventObject {
  account: string;
}
export type PausedEvent = TypedEvent<[string], PausedEventObject>;

export type PausedEventFilter = TypedEventFilter<PausedEvent>;

export interface TransactionDepositedEventObject {
  from: string;
  to: string;
  version: BigNumber;
  opaqueData: string;
}
export type TransactionDepositedEvent = TypedEvent<
  [string, string, BigNumber, string],
  TransactionDepositedEventObject
>;

export type TransactionDepositedEventFilter =
  TypedEventFilter<TransactionDepositedEvent>;

export interface UnpausedEventObject {
  account: string;
}
export type UnpausedEvent = TypedEvent<[string], UnpausedEventObject>;

export type UnpausedEventFilter = TypedEventFilter<UnpausedEvent>;

export interface WithdrawalFinalizedEventObject {
  withdrawalHash: string;
  success: boolean;
}
export type WithdrawalFinalizedEvent = TypedEvent<
  [string, boolean],
  WithdrawalFinalizedEventObject
>;

export type WithdrawalFinalizedEventFilter =
  TypedEventFilter<WithdrawalFinalizedEvent>;

export interface WithdrawalProvenEventObject {
  withdrawalHash: string;
  from: string;
  to: string;
}
export type WithdrawalProvenEvent = TypedEvent<
  [string, string, string],
  WithdrawalProvenEventObject
>;

export type WithdrawalProvenEventFilter =
  TypedEventFilter<WithdrawalProvenEvent>;

export interface OptimismPortal extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: OptimismPortalInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    BASE_FEE_MAX_CHANGE_DENOMINATOR(
      overrides?: CallOverrides
    ): Promise<[BigNumber]>;

    ELASTICITY_MULTIPLIER(overrides?: CallOverrides): Promise<[BigNumber]>;

    FINALIZATION_PERIOD_SECONDS(
      overrides?: CallOverrides
    ): Promise<[BigNumber]>;

    GUARDIAN(overrides?: CallOverrides): Promise<[string]>;

    INITIAL_BASE_FEE(overrides?: CallOverrides): Promise<[BigNumber]>;

    L2_ORACLE(overrides?: CallOverrides): Promise<[string]>;

    MAXIMUM_BASE_FEE(overrides?: CallOverrides): Promise<[BigNumber]>;

    MAX_RESOURCE_LIMIT(overrides?: CallOverrides): Promise<[BigNumber]>;

    MINIMUM_BASE_FEE(overrides?: CallOverrides): Promise<[BigNumber]>;

    TARGET_RESOURCE_LIMIT(overrides?: CallOverrides): Promise<[BigNumber]>;

    depositTransaction(
      _to: PromiseOrValue<string>,
      _value: PromiseOrValue<BigNumberish>,
      _gasLimit: PromiseOrValue<BigNumberish>,
      _isCreation: PromiseOrValue<boolean>,
      _data: PromiseOrValue<BytesLike>,
      overrides?: PayableOverrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    donateETH(
      overrides?: PayableOverrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    finalizeWithdrawalTransaction(
      _tx: Types.WithdrawalTransactionStruct,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    finalizedWithdrawals(
      arg0: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<[boolean]>;

    initialize(
      _paused: PromiseOrValue<boolean>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    isOutputFinalized(
      _l2OutputIndex: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<[boolean]>;

    l2Sender(overrides?: CallOverrides): Promise<[string]>;

    params(
      overrides?: CallOverrides
    ): Promise<
      [BigNumber, BigNumber, BigNumber] & {
        prevBaseFee: BigNumber;
        prevBoughtGas: BigNumber;
        prevBlockNum: BigNumber;
      }
    >;

    pause(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    paused(overrides?: CallOverrides): Promise<[boolean]>;

    proveWithdrawalTransaction(
      _tx: Types.WithdrawalTransactionStruct,
      _l2OutputIndex: PromiseOrValue<BigNumberish>,
      _outputRootProof: Types.OutputRootProofStruct,
      _withdrawalProof: PromiseOrValue<BytesLike>[],
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    provenWithdrawals(
      arg0: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<
      [string, BigNumber, BigNumber] & {
        outputRoot: string;
        timestamp: BigNumber;
        l2OutputIndex: BigNumber;
      }
    >;

    unpause(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    version(overrides?: CallOverrides): Promise<[string]>;
  };

  BASE_FEE_MAX_CHANGE_DENOMINATOR(
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  ELASTICITY_MULTIPLIER(overrides?: CallOverrides): Promise<BigNumber>;

  FINALIZATION_PERIOD_SECONDS(overrides?: CallOverrides): Promise<BigNumber>;

  GUARDIAN(overrides?: CallOverrides): Promise<string>;

  INITIAL_BASE_FEE(overrides?: CallOverrides): Promise<BigNumber>;

  L2_ORACLE(overrides?: CallOverrides): Promise<string>;

  MAXIMUM_BASE_FEE(overrides?: CallOverrides): Promise<BigNumber>;

  MAX_RESOURCE_LIMIT(overrides?: CallOverrides): Promise<BigNumber>;

  MINIMUM_BASE_FEE(overrides?: CallOverrides): Promise<BigNumber>;

  TARGET_RESOURCE_LIMIT(overrides?: CallOverrides): Promise<BigNumber>;

  depositTransaction(
    _to: PromiseOrValue<string>,
    _value: PromiseOrValue<BigNumberish>,
    _gasLimit: PromiseOrValue<BigNumberish>,
    _isCreation: PromiseOrValue<boolean>,
    _data: PromiseOrValue<BytesLike>,
    overrides?: PayableOverrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  donateETH(
    overrides?: PayableOverrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  finalizeWithdrawalTransaction(
    _tx: Types.WithdrawalTransactionStruct,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  finalizedWithdrawals(
    arg0: PromiseOrValue<BytesLike>,
    overrides?: CallOverrides
  ): Promise<boolean>;

  initialize(
    _paused: PromiseOrValue<boolean>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  isOutputFinalized(
    _l2OutputIndex: PromiseOrValue<BigNumberish>,
    overrides?: CallOverrides
  ): Promise<boolean>;

  l2Sender(overrides?: CallOverrides): Promise<string>;

  params(
    overrides?: CallOverrides
  ): Promise<
    [BigNumber, BigNumber, BigNumber] & {
      prevBaseFee: BigNumber;
      prevBoughtGas: BigNumber;
      prevBlockNum: BigNumber;
    }
  >;

  pause(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  paused(overrides?: CallOverrides): Promise<boolean>;

  proveWithdrawalTransaction(
    _tx: Types.WithdrawalTransactionStruct,
    _l2OutputIndex: PromiseOrValue<BigNumberish>,
    _outputRootProof: Types.OutputRootProofStruct,
    _withdrawalProof: PromiseOrValue<BytesLike>[],
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  provenWithdrawals(
    arg0: PromiseOrValue<BytesLike>,
    overrides?: CallOverrides
  ): Promise<
    [string, BigNumber, BigNumber] & {
      outputRoot: string;
      timestamp: BigNumber;
      l2OutputIndex: BigNumber;
    }
  >;

  unpause(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  version(overrides?: CallOverrides): Promise<string>;

  callStatic: {
    BASE_FEE_MAX_CHANGE_DENOMINATOR(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    ELASTICITY_MULTIPLIER(overrides?: CallOverrides): Promise<BigNumber>;

    FINALIZATION_PERIOD_SECONDS(overrides?: CallOverrides): Promise<BigNumber>;

    GUARDIAN(overrides?: CallOverrides): Promise<string>;

    INITIAL_BASE_FEE(overrides?: CallOverrides): Promise<BigNumber>;

    L2_ORACLE(overrides?: CallOverrides): Promise<string>;

    MAXIMUM_BASE_FEE(overrides?: CallOverrides): Promise<BigNumber>;

    MAX_RESOURCE_LIMIT(overrides?: CallOverrides): Promise<BigNumber>;

    MINIMUM_BASE_FEE(overrides?: CallOverrides): Promise<BigNumber>;

    TARGET_RESOURCE_LIMIT(overrides?: CallOverrides): Promise<BigNumber>;

    depositTransaction(
      _to: PromiseOrValue<string>,
      _value: PromiseOrValue<BigNumberish>,
      _gasLimit: PromiseOrValue<BigNumberish>,
      _isCreation: PromiseOrValue<boolean>,
      _data: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<void>;

    donateETH(overrides?: CallOverrides): Promise<void>;

    finalizeWithdrawalTransaction(
      _tx: Types.WithdrawalTransactionStruct,
      overrides?: CallOverrides
    ): Promise<void>;

    finalizedWithdrawals(
      arg0: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<boolean>;

    initialize(
      _paused: PromiseOrValue<boolean>,
      overrides?: CallOverrides
    ): Promise<void>;

    isOutputFinalized(
      _l2OutputIndex: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<boolean>;

    l2Sender(overrides?: CallOverrides): Promise<string>;

    params(
      overrides?: CallOverrides
    ): Promise<
      [BigNumber, BigNumber, BigNumber] & {
        prevBaseFee: BigNumber;
        prevBoughtGas: BigNumber;
        prevBlockNum: BigNumber;
      }
    >;

    pause(overrides?: CallOverrides): Promise<void>;

    paused(overrides?: CallOverrides): Promise<boolean>;

    proveWithdrawalTransaction(
      _tx: Types.WithdrawalTransactionStruct,
      _l2OutputIndex: PromiseOrValue<BigNumberish>,
      _outputRootProof: Types.OutputRootProofStruct,
      _withdrawalProof: PromiseOrValue<BytesLike>[],
      overrides?: CallOverrides
    ): Promise<void>;

    provenWithdrawals(
      arg0: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<
      [string, BigNumber, BigNumber] & {
        outputRoot: string;
        timestamp: BigNumber;
        l2OutputIndex: BigNumber;
      }
    >;

    unpause(overrides?: CallOverrides): Promise<void>;

    version(overrides?: CallOverrides): Promise<string>;
  };

  filters: {
    "Initialized(uint8)"(version?: null): InitializedEventFilter;
    Initialized(version?: null): InitializedEventFilter;

    "Paused(address)"(account?: null): PausedEventFilter;
    Paused(account?: null): PausedEventFilter;

    "TransactionDeposited(address,address,uint256,bytes)"(
      from?: PromiseOrValue<string> | null,
      to?: PromiseOrValue<string> | null,
      version?: PromiseOrValue<BigNumberish> | null,
      opaqueData?: null
    ): TransactionDepositedEventFilter;
    TransactionDeposited(
      from?: PromiseOrValue<string> | null,
      to?: PromiseOrValue<string> | null,
      version?: PromiseOrValue<BigNumberish> | null,
      opaqueData?: null
    ): TransactionDepositedEventFilter;

    "Unpaused(address)"(account?: null): UnpausedEventFilter;
    Unpaused(account?: null): UnpausedEventFilter;

    "WithdrawalFinalized(bytes32,bool)"(
      withdrawalHash?: PromiseOrValue<BytesLike> | null,
      success?: null
    ): WithdrawalFinalizedEventFilter;
    WithdrawalFinalized(
      withdrawalHash?: PromiseOrValue<BytesLike> | null,
      success?: null
    ): WithdrawalFinalizedEventFilter;

    "WithdrawalProven(bytes32,address,address)"(
      withdrawalHash?: PromiseOrValue<BytesLike> | null,
      from?: PromiseOrValue<string> | null,
      to?: PromiseOrValue<string> | null
    ): WithdrawalProvenEventFilter;
    WithdrawalProven(
      withdrawalHash?: PromiseOrValue<BytesLike> | null,
      from?: PromiseOrValue<string> | null,
      to?: PromiseOrValue<string> | null
    ): WithdrawalProvenEventFilter;
  };

  estimateGas: {
    BASE_FEE_MAX_CHANGE_DENOMINATOR(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    ELASTICITY_MULTIPLIER(overrides?: CallOverrides): Promise<BigNumber>;

    FINALIZATION_PERIOD_SECONDS(overrides?: CallOverrides): Promise<BigNumber>;

    GUARDIAN(overrides?: CallOverrides): Promise<BigNumber>;

    INITIAL_BASE_FEE(overrides?: CallOverrides): Promise<BigNumber>;

    L2_ORACLE(overrides?: CallOverrides): Promise<BigNumber>;

    MAXIMUM_BASE_FEE(overrides?: CallOverrides): Promise<BigNumber>;

    MAX_RESOURCE_LIMIT(overrides?: CallOverrides): Promise<BigNumber>;

    MINIMUM_BASE_FEE(overrides?: CallOverrides): Promise<BigNumber>;

    TARGET_RESOURCE_LIMIT(overrides?: CallOverrides): Promise<BigNumber>;

    depositTransaction(
      _to: PromiseOrValue<string>,
      _value: PromiseOrValue<BigNumberish>,
      _gasLimit: PromiseOrValue<BigNumberish>,
      _isCreation: PromiseOrValue<boolean>,
      _data: PromiseOrValue<BytesLike>,
      overrides?: PayableOverrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    donateETH(
      overrides?: PayableOverrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    finalizeWithdrawalTransaction(
      _tx: Types.WithdrawalTransactionStruct,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    finalizedWithdrawals(
      arg0: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    initialize(
      _paused: PromiseOrValue<boolean>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    isOutputFinalized(
      _l2OutputIndex: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    l2Sender(overrides?: CallOverrides): Promise<BigNumber>;

    params(overrides?: CallOverrides): Promise<BigNumber>;

    pause(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    paused(overrides?: CallOverrides): Promise<BigNumber>;

    proveWithdrawalTransaction(
      _tx: Types.WithdrawalTransactionStruct,
      _l2OutputIndex: PromiseOrValue<BigNumberish>,
      _outputRootProof: Types.OutputRootProofStruct,
      _withdrawalProof: PromiseOrValue<BytesLike>[],
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    provenWithdrawals(
      arg0: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    unpause(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    version(overrides?: CallOverrides): Promise<BigNumber>;
  };

  populateTransaction: {
    BASE_FEE_MAX_CHANGE_DENOMINATOR(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    ELASTICITY_MULTIPLIER(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    FINALIZATION_PERIOD_SECONDS(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    GUARDIAN(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    INITIAL_BASE_FEE(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    L2_ORACLE(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    MAXIMUM_BASE_FEE(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    MAX_RESOURCE_LIMIT(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    MINIMUM_BASE_FEE(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    TARGET_RESOURCE_LIMIT(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    depositTransaction(
      _to: PromiseOrValue<string>,
      _value: PromiseOrValue<BigNumberish>,
      _gasLimit: PromiseOrValue<BigNumberish>,
      _isCreation: PromiseOrValue<boolean>,
      _data: PromiseOrValue<BytesLike>,
      overrides?: PayableOverrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    donateETH(
      overrides?: PayableOverrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    finalizeWithdrawalTransaction(
      _tx: Types.WithdrawalTransactionStruct,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    finalizedWithdrawals(
      arg0: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    initialize(
      _paused: PromiseOrValue<boolean>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    isOutputFinalized(
      _l2OutputIndex: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    l2Sender(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    params(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    pause(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    paused(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    proveWithdrawalTransaction(
      _tx: Types.WithdrawalTransactionStruct,
      _l2OutputIndex: PromiseOrValue<BigNumberish>,
      _outputRootProof: Types.OutputRootProofStruct,
      _withdrawalProof: PromiseOrValue<BytesLike>[],
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    provenWithdrawals(
      arg0: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    unpause(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    version(overrides?: CallOverrides): Promise<PopulatedTransaction>;
  };
}
