// @eth-optimism/contracts-bedrock/dist/types/contracts/echidna/FuzzOptimismPortal.sol/EchidnaFuzzOptimismPortal.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  PayableOverrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../../common";

export interface EchidnaFuzzOptimismPortalInterface extends utils.Interface {
  functions: {
    "echidna_deposit_completes()": FunctionFragment;
    "testDepositTransactionCompletes(address,uint256,uint256,uint64,bool,bytes)": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "echidna_deposit_completes"
      | "testDepositTransactionCompletes"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "echidna_deposit_completes",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "testDepositTransactionCompletes",
    values: [
      PromiseOrValue<string>,
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<boolean>,
      PromiseOrValue<BytesLike>
    ]
  ): string;

  decodeFunctionResult(
    functionFragment: "echidna_deposit_completes",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "testDepositTransactionCompletes",
    data: BytesLike
  ): Result;

  events: {};
}

export interface EchidnaFuzzOptimismPortal extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: EchidnaFuzzOptimismPortalInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    echidna_deposit_completes(overrides?: CallOverrides): Promise<[boolean]>;

    testDepositTransactionCompletes(
      _to: PromiseOrValue<string>,
      _mint: PromiseOrValue<BigNumberish>,
      _value: PromiseOrValue<BigNumberish>,
      _gasLimit: PromiseOrValue<BigNumberish>,
      _isCreation: PromiseOrValue<boolean>,
      _data: PromiseOrValue<BytesLike>,
      overrides?: PayableOverrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;
  };

  echidna_deposit_completes(overrides?: CallOverrides): Promise<boolean>;

  testDepositTransactionCompletes(
    _to: PromiseOrValue<string>,
    _mint: PromiseOrValue<BigNumberish>,
    _value: PromiseOrValue<BigNumberish>,
    _gasLimit: PromiseOrValue<BigNumberish>,
    _isCreation: PromiseOrValue<boolean>,
    _data: PromiseOrValue<BytesLike>,
    overrides?: PayableOverrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    echidna_deposit_completes(overrides?: CallOverrides): Promise<boolean>;

    testDepositTransactionCompletes(
      _to: PromiseOrValue<string>,
      _mint: PromiseOrValue<BigNumberish>,
      _value: PromiseOrValue<BigNumberish>,
      _gasLimit: PromiseOrValue<BigNumberish>,
      _isCreation: PromiseOrValue<boolean>,
      _data: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {};

  estimateGas: {
    echidna_deposit_completes(overrides?: CallOverrides): Promise<BigNumber>;

    testDepositTransactionCompletes(
      _to: PromiseOrValue<string>,
      _mint: PromiseOrValue<BigNumberish>,
      _value: PromiseOrValue<BigNumberish>,
      _gasLimit: PromiseOrValue<BigNumberish>,
      _isCreation: PromiseOrValue<boolean>,
      _data: PromiseOrValue<BytesLike>,
      overrides?: PayableOverrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    echidna_deposit_completes(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    testDepositTransactionCompletes(
      _to: PromiseOrValue<string>,
      _mint: PromiseOrValue<BigNumberish>,
      _value: PromiseOrValue<BigNumberish>,
      _gasLimit: PromiseOrValue<BigNumberish>,
      _isCreation: PromiseOrValue<boolean>,
      _data: PromiseOrValue<BytesLike>,
      overrides?: PayableOverrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;
  };
}
