// @eth-optimism/contracts-bedrock/dist/types/contracts/test/invariants/Encoding.t.sol/Encoding_Converter.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../../../common";

export interface Encoding_ConverterInterface extends utils.Interface {
  functions: {
    "convertRoundTripAToB(uint240,uint16)": FunctionFragment;
    "convertRoundTripBToA(uint256)": FunctionFragment;
    "failedRoundtripAToB()": FunctionFragment;
    "failedRoundtripBToA()": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "convertRoundTripAToB"
      | "convertRoundTripBToA"
      | "failedRoundtripAToB"
      | "failedRoundtripBToA"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "convertRoundTripAToB",
    values: [PromiseOrValue<BigNumberish>, PromiseOrValue<BigNumberish>]
  ): string;
  encodeFunctionData(
    functionFragment: "convertRoundTripBToA",
    values: [PromiseOrValue<BigNumberish>]
  ): string;
  encodeFunctionData(
    functionFragment: "failedRoundtripAToB",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "failedRoundtripBToA",
    values?: undefined
  ): string;

  decodeFunctionResult(
    functionFragment: "convertRoundTripAToB",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "convertRoundTripBToA",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "failedRoundtripAToB",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "failedRoundtripBToA",
    data: BytesLike
  ): Result;

  events: {};
}

export interface Encoding_Converter extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: Encoding_ConverterInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    convertRoundTripAToB(
      _nonce: PromiseOrValue<BigNumberish>,
      _version: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    convertRoundTripBToA(
      _versionedNonce: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    failedRoundtripAToB(overrides?: CallOverrides): Promise<[boolean]>;

    failedRoundtripBToA(overrides?: CallOverrides): Promise<[boolean]>;
  };

  convertRoundTripAToB(
    _nonce: PromiseOrValue<BigNumberish>,
    _version: PromiseOrValue<BigNumberish>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  convertRoundTripBToA(
    _versionedNonce: PromiseOrValue<BigNumberish>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  failedRoundtripAToB(overrides?: CallOverrides): Promise<boolean>;

  failedRoundtripBToA(overrides?: CallOverrides): Promise<boolean>;

  callStatic: {
    convertRoundTripAToB(
      _nonce: PromiseOrValue<BigNumberish>,
      _version: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<void>;

    convertRoundTripBToA(
      _versionedNonce: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<void>;

    failedRoundtripAToB(overrides?: CallOverrides): Promise<boolean>;

    failedRoundtripBToA(overrides?: CallOverrides): Promise<boolean>;
  };

  filters: {};

  estimateGas: {
    convertRoundTripAToB(
      _nonce: PromiseOrValue<BigNumberish>,
      _version: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    convertRoundTripBToA(
      _versionedNonce: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    failedRoundtripAToB(overrides?: CallOverrides): Promise<BigNumber>;

    failedRoundtripBToA(overrides?: CallOverrides): Promise<BigNumber>;
  };

  populateTransaction: {
    convertRoundTripAToB(
      _nonce: PromiseOrValue<BigNumberish>,
      _version: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    convertRoundTripBToA(
      _versionedNonce: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    failedRoundtripAToB(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    failedRoundtripBToA(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;
  };
}
