// @wagmi/core/dist/index.js
import "./chunk-KX4UEHS5.js";
import {
  AddChainError,
  ChainDoesNotSupportMulticallError,
  ChainMismatchError,
  ChainNotConfiguredError,
  Client,
  ConnectorAlreadyConnectedError,
  ConnectorNotFoundError,
  ContractMethodDoesNotExistError,
  ContractMethodNoResultError,
  ContractMethodRevertedError,
  ContractResultDecodeError,
  ProviderChainsNotFound,
  ProviderRpcError,
  ResourceUnavailableError,
  RpcError,
  SwitchChainError,
  SwitchChainNotSupportedError,
  UserRejectedRequestError,
  configureChains,
  connect,
  createClient,
  createStorage,
  deepEqual,
  deserialize,
  disconnect,
  erc20ABI,
  erc4626ABI,
  erc721ABI,
  fetchBalance,
  fetchBlockNumber,
  fetchEnsAddress,
  fetchEnsAvatar,
  fetchEnsName,
  fetchEnsResolver,
  fetchFeeData,
  fetchSigner,
  fetchToken,
  fetchTransaction,
  getAccount,
  getClient,
  getContract,
  getNetwork,
  getProvider,
  getWebSocketProvider,
  minimizeContractInterface,
  multicall,
  noopStorage,
  normalizeChainId,
  parseContractResult,
  prepareSendTransaction,
  prepareWriteContract,
  readContract,
  readContracts,
  sendTransaction,
  serialize,
  signMessage,
  signTypedData,
  switchNetwork,
  units,
  waitForTransaction,
  watchAccount,
  watchBlockNumber,
  watchContractEvent,
  watchMulticall,
  watchNetwork,
  watchPendingTransactions,
  watchProvider,
  watchReadContract,
  watchReadContracts,
  watchSigner,
  watchWebSocketProvider,
  writeContract
} from "./chunk-55IO54NW.js";
import {
  goerli,
  mainnet
} from "./chunk-FCJXFFXS.js";
import {
  Connector
} from "./chunk-BVC4KGLQ.js";
import {
  InjectedConnector
} from "./chunk-EQOEZP46.js";
import "./chunk-MQXBDTVK.js";
export {
  AddChainError,
  ChainDoesNotSupportMulticallError,
  ChainMismatchError,
  ChainNotConfiguredError,
  Client,
  Connector,
  ConnectorAlreadyConnectedError,
  ConnectorNotFoundError,
  ContractMethodDoesNotExistError,
  ContractMethodNoResultError,
  ContractMethodRevertedError,
  ContractResultDecodeError,
  InjectedConnector,
  ProviderChainsNotFound,
  ProviderRpcError,
  ResourceUnavailableError,
  RpcError,
  SwitchChainError,
  SwitchChainNotSupportedError,
  UserRejectedRequestError,
  configureChains,
  connect,
  createClient,
  createStorage,
  deepEqual,
  deserialize,
  disconnect,
  erc20ABI,
  erc4626ABI,
  erc721ABI,
  fetchBalance,
  fetchBlockNumber,
  fetchEnsAddress,
  fetchEnsAvatar,
  fetchEnsName,
  fetchEnsResolver,
  fetchFeeData,
  fetchSigner,
  fetchToken,
  fetchTransaction,
  getAccount,
  getClient,
  getContract,
  getNetwork,
  getProvider,
  getWebSocketProvider,
  goerli,
  mainnet,
  minimizeContractInterface,
  multicall,
  noopStorage,
  normalizeChainId,
  parseContractResult,
  prepareSendTransaction,
  prepareWriteContract,
  readContract,
  readContracts,
  sendTransaction,
  serialize,
  signMessage,
  signTypedData,
  switchNetwork,
  units,
  waitForTransaction,
  watchAccount,
  watchBlockNumber,
  watchContractEvent,
  watchMulticall,
  watchNetwork,
  watchPendingTransactions,
  watchProvider,
  watchReadContract,
  watchReadContracts,
  watchSigner,
  watchWebSocketProvider,
  writeContract
};
