// @eth-optimism/contracts-bedrock/dist/types/contracts/test/invariants/Burn.Gas.t.sol/Burn_GasBurner.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../../../common";

export interface Burn_GasBurnerInterface extends utils.Interface {
  functions: {
    "burnGas(uint256)": FunctionFragment;
    "failedGasBurn()": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic: "burnGas" | "failedGasBurn"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "burnGas",
    values: [PromiseOrValue<BigNumberish>]
  ): string;
  encodeFunctionData(
    functionFragment: "failedGasBurn",
    values?: undefined
  ): string;

  decodeFunctionResult(functionFragment: "burnGas", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "failedGasBurn",
    data: BytesLike
  ): Result;

  events: {};
}

export interface Burn_GasBurner extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: Burn_GasBurnerInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    burnGas(
      _value: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    failedGasBurn(overrides?: CallOverrides): Promise<[boolean]>;
  };

  burnGas(
    _value: PromiseOrValue<BigNumberish>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  failedGasBurn(overrides?: CallOverrides): Promise<boolean>;

  callStatic: {
    burnGas(
      _value: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<void>;

    failedGasBurn(overrides?: CallOverrides): Promise<boolean>;
  };

  filters: {};

  estimateGas: {
    burnGas(
      _value: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    failedGasBurn(overrides?: CallOverrides): Promise<BigNumber>;
  };

  populateTransaction: {
    burnGas(
      _value: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    failedGasBurn(overrides?: CallOverrides): Promise<PopulatedTransaction>;
  };
}
