// @eth-optimism/contracts-bedrock/dist/types/contracts/L2/SequencerFeeVault.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type {
  FunctionFragment,
  Result,
  EventFragment,
} from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../common";

export interface SequencerFeeVaultInterface extends utils.Interface {
  functions: {
    "MIN_WITHDRAWAL_AMOUNT()": FunctionFragment;
    "RECIPIENT()": FunctionFragment;
    "l1FeeWallet()": FunctionFragment;
    "totalProcessed()": FunctionFragment;
    "version()": FunctionFragment;
    "withdraw()": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "MIN_WITHDRAWAL_AMOUNT"
      | "RECIPIENT"
      | "l1FeeWallet"
      | "totalProcessed"
      | "version"
      | "withdraw"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "MIN_WITHDRAWAL_AMOUNT",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "RECIPIENT", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "l1FeeWallet",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "totalProcessed",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "version", values?: undefined): string;
  encodeFunctionData(functionFragment: "withdraw", values?: undefined): string;

  decodeFunctionResult(
    functionFragment: "MIN_WITHDRAWAL_AMOUNT",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "RECIPIENT", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "l1FeeWallet",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "totalProcessed",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "version", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "withdraw", data: BytesLike): Result;

  events: {
    "Withdrawal(uint256,address,address)": EventFragment;
  };

  getEvent(nameOrSignatureOrTopic: "Withdrawal"): EventFragment;
}

export interface WithdrawalEventObject {
  value: BigNumber;
  to: string;
  from: string;
}
export type WithdrawalEvent = TypedEvent<
  [BigNumber, string, string],
  WithdrawalEventObject
>;

export type WithdrawalEventFilter = TypedEventFilter<WithdrawalEvent>;

export interface SequencerFeeVault extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: SequencerFeeVaultInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    MIN_WITHDRAWAL_AMOUNT(overrides?: CallOverrides): Promise<[BigNumber]>;

    RECIPIENT(overrides?: CallOverrides): Promise<[string]>;

    l1FeeWallet(overrides?: CallOverrides): Promise<[string]>;

    totalProcessed(overrides?: CallOverrides): Promise<[BigNumber]>;

    version(overrides?: CallOverrides): Promise<[string]>;

    withdraw(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;
  };

  MIN_WITHDRAWAL_AMOUNT(overrides?: CallOverrides): Promise<BigNumber>;

  RECIPIENT(overrides?: CallOverrides): Promise<string>;

  l1FeeWallet(overrides?: CallOverrides): Promise<string>;

  totalProcessed(overrides?: CallOverrides): Promise<BigNumber>;

  version(overrides?: CallOverrides): Promise<string>;

  withdraw(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    MIN_WITHDRAWAL_AMOUNT(overrides?: CallOverrides): Promise<BigNumber>;

    RECIPIENT(overrides?: CallOverrides): Promise<string>;

    l1FeeWallet(overrides?: CallOverrides): Promise<string>;

    totalProcessed(overrides?: CallOverrides): Promise<BigNumber>;

    version(overrides?: CallOverrides): Promise<string>;

    withdraw(overrides?: CallOverrides): Promise<void>;
  };

  filters: {
    "Withdrawal(uint256,address,address)"(
      value?: null,
      to?: null,
      from?: null
    ): WithdrawalEventFilter;
    Withdrawal(value?: null, to?: null, from?: null): WithdrawalEventFilter;
  };

  estimateGas: {
    MIN_WITHDRAWAL_AMOUNT(overrides?: CallOverrides): Promise<BigNumber>;

    RECIPIENT(overrides?: CallOverrides): Promise<BigNumber>;

    l1FeeWallet(overrides?: CallOverrides): Promise<BigNumber>;

    totalProcessed(overrides?: CallOverrides): Promise<BigNumber>;

    version(overrides?: CallOverrides): Promise<BigNumber>;

    withdraw(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    MIN_WITHDRAWAL_AMOUNT(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    RECIPIENT(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    l1FeeWallet(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    totalProcessed(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    version(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    withdraw(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;
  };
}
