// @eth-optimism/contracts-bedrock/dist/types/factories/contracts/L2/CrossDomainOwnable2__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  CrossDomainOwnable2,
  CrossDomainOwnable2Interface,
} from "../../../contracts/L2/CrossDomainOwnable2";

const _abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "previousOwner",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "newOwner",
        type: "address",
      },
    ],
    name: "OwnershipTransferred",
    type: "event",
  },
  {
    inputs: [],
    name: "owner",
    outputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "renounceOwnership",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "newOwner",
        type: "address",
      },
    ],
    name: "transferOwnership",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
];

export class CrossDomainOwnable2__factory {
  static readonly abi = _abi;
  static createInterface(): CrossDomainOwnable2Interface {
    return new utils.Interface(_abi) as CrossDomainOwnable2Interface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): CrossDomainOwnable2 {
    return new Contract(address, _abi, signerOrProvider) as CrossDomainOwnable2;
  }
}
