// @eth-optimism/contracts-bedrock/dist/types/contracts/test/invariants/AddressAliasHelper.t.sol/AddressAliasHelper_Converter.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../../../common";

export interface AddressAliasHelper_ConverterInterface extends utils.Interface {
  functions: {
    "convertRoundTrip(address)": FunctionFragment;
    "failedRoundtrip()": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic: "convertRoundTrip" | "failedRoundtrip"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "convertRoundTrip",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "failedRoundtrip",
    values?: undefined
  ): string;

  decodeFunctionResult(
    functionFragment: "convertRoundTrip",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "failedRoundtrip",
    data: BytesLike
  ): Result;

  events: {};
}

export interface AddressAliasHelper_Converter extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: AddressAliasHelper_ConverterInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    convertRoundTrip(
      addr: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    failedRoundtrip(overrides?: CallOverrides): Promise<[boolean]>;
  };

  convertRoundTrip(
    addr: PromiseOrValue<string>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  failedRoundtrip(overrides?: CallOverrides): Promise<boolean>;

  callStatic: {
    convertRoundTrip(
      addr: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<void>;

    failedRoundtrip(overrides?: CallOverrides): Promise<boolean>;
  };

  filters: {};

  estimateGas: {
    convertRoundTrip(
      addr: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    failedRoundtrip(overrides?: CallOverrides): Promise<BigNumber>;
  };

  populateTransaction: {
    convertRoundTrip(
      addr: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    failedRoundtrip(overrides?: CallOverrides): Promise<PopulatedTransaction>;
  };
}
