// @eth-optimism/contracts-bedrock/dist/types/contracts/test/RLPReader.t.sol/RLPReader_readList_Test.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type {
  FunctionFragment,
  Result,
  EventFragment,
} from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../../common";

export declare namespace StdInvariant {
  export type FuzzSelectorStruct = {
    addr: PromiseOrValue<string>;
    selectors: PromiseOrValue<BytesLike>[];
  };

  export type FuzzSelectorStructOutput = [string, string[]] & {
    addr: string;
    selectors: string[];
  };
}

export interface RLPReader_readList_TestInterface extends utils.Interface {
  functions: {
    "IS_TEST()": FunctionFragment;
    "excludeArtifacts()": FunctionFragment;
    "excludeContracts()": FunctionFragment;
    "excludeSenders()": FunctionFragment;
    "failed()": FunctionFragment;
    "setUp()": FunctionFragment;
    "targetArtifactSelectors()": FunctionFragment;
    "targetArtifacts()": FunctionFragment;
    "targetContracts()": FunctionFragment;
    "targetSelectors()": FunctionFragment;
    "targetSenders()": FunctionFragment;
    "test_readList_dictTest1_succeeds()": FunctionFragment;
    "test_readList_empty_succeeds()": FunctionFragment;
    "test_readList_incorrectLengthInArray_reverts()": FunctionFragment;
    "test_readList_int32Overflow2_reverts()": FunctionFragment;
    "test_readList_int32Overflow_reverts()": FunctionFragment;
    "test_readList_invalidRemainder_reverts()": FunctionFragment;
    "test_readList_invalidShortList_reverts()": FunctionFragment;
    "test_readList_invalidValue_reverts()": FunctionFragment;
    "test_readList_leadingZerosInLongLengthArray1_reverts()": FunctionFragment;
    "test_readList_leadingZerosInLongLengthArray2_reverts()": FunctionFragment;
    "test_readList_leadingZerosInLongLengthList1_reverts()": FunctionFragment;
    "test_readList_listLongerThan32Elements_reverts()": FunctionFragment;
    "test_readList_listOfLists2_succeeds()": FunctionFragment;
    "test_readList_listOfLists_succeeds()": FunctionFragment;
    "test_readList_longList1_succeeds()": FunctionFragment;
    "test_readList_longList2_succeeds()": FunctionFragment;
    "test_readList_longListLessThan56Bytes_reverts()": FunctionFragment;
    "test_readList_longStringLength_reverts()": FunctionFragment;
    "test_readList_longStringLessThan56Bytes_reverts()": FunctionFragment;
    "test_readList_multiList_succeeds()": FunctionFragment;
    "test_readList_nonOptimalLongLengthArray1_reverts()": FunctionFragment;
    "test_readList_nonOptimalLongLengthArray2_reverts()": FunctionFragment;
    "test_readList_notEnoughContentForList1_reverts()": FunctionFragment;
    "test_readList_notEnoughContentForList2_reverts()": FunctionFragment;
    "test_readList_notEnoughContentForString1_reverts()": FunctionFragment;
    "test_readList_notEnoughContentForString2_reverts()": FunctionFragment;
    "test_readList_notLongEnough_reverts()": FunctionFragment;
    "test_readList_shortListMax1_succeeds()": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "IS_TEST"
      | "excludeArtifacts"
      | "excludeContracts"
      | "excludeSenders"
      | "failed"
      | "setUp"
      | "targetArtifactSelectors"
      | "targetArtifacts"
      | "targetContracts"
      | "targetSelectors"
      | "targetSenders"
      | "test_readList_dictTest1_succeeds"
      | "test_readList_empty_succeeds"
      | "test_readList_incorrectLengthInArray_reverts"
      | "test_readList_int32Overflow2_reverts"
      | "test_readList_int32Overflow_reverts"
      | "test_readList_invalidRemainder_reverts"
      | "test_readList_invalidShortList_reverts"
      | "test_readList_invalidValue_reverts"
      | "test_readList_leadingZerosInLongLengthArray1_reverts"
      | "test_readList_leadingZerosInLongLengthArray2_reverts"
      | "test_readList_leadingZerosInLongLengthList1_reverts"
      | "test_readList_listLongerThan32Elements_reverts"
      | "test_readList_listOfLists2_succeeds"
      | "test_readList_listOfLists_succeeds"
      | "test_readList_longList1_succeeds"
      | "test_readList_longList2_succeeds"
      | "test_readList_longListLessThan56Bytes_reverts"
      | "test_readList_longStringLength_reverts"
      | "test_readList_longStringLessThan56Bytes_reverts"
      | "test_readList_multiList_succeeds"
      | "test_readList_nonOptimalLongLengthArray1_reverts"
      | "test_readList_nonOptimalLongLengthArray2_reverts"
      | "test_readList_notEnoughContentForList1_reverts"
      | "test_readList_notEnoughContentForList2_reverts"
      | "test_readList_notEnoughContentForString1_reverts"
      | "test_readList_notEnoughContentForString2_reverts"
      | "test_readList_notLongEnough_reverts"
      | "test_readList_shortListMax1_succeeds"
  ): FunctionFragment;

  encodeFunctionData(functionFragment: "IS_TEST", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "excludeArtifacts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "excludeContracts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "excludeSenders",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "failed", values?: undefined): string;
  encodeFunctionData(functionFragment: "setUp", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "targetArtifactSelectors",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "targetArtifacts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "targetContracts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "targetSelectors",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "targetSenders",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_dictTest1_succeeds",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_empty_succeeds",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_incorrectLengthInArray_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_int32Overflow2_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_int32Overflow_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_invalidRemainder_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_invalidShortList_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_invalidValue_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_leadingZerosInLongLengthArray1_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_leadingZerosInLongLengthArray2_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_leadingZerosInLongLengthList1_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_listLongerThan32Elements_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_listOfLists2_succeeds",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_listOfLists_succeeds",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_longList1_succeeds",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_longList2_succeeds",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_longListLessThan56Bytes_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_longStringLength_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_longStringLessThan56Bytes_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_multiList_succeeds",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_nonOptimalLongLengthArray1_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_nonOptimalLongLengthArray2_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_notEnoughContentForList1_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_notEnoughContentForList2_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_notEnoughContentForString1_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_notEnoughContentForString2_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_notLongEnough_reverts",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "test_readList_shortListMax1_succeeds",
    values?: undefined
  ): string;

  decodeFunctionResult(functionFragment: "IS_TEST", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "excludeArtifacts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "excludeContracts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "excludeSenders",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "failed", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "setUp", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "targetArtifactSelectors",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "targetArtifacts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "targetContracts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "targetSelectors",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "targetSenders",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_dictTest1_succeeds",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_empty_succeeds",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_incorrectLengthInArray_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_int32Overflow2_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_int32Overflow_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_invalidRemainder_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_invalidShortList_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_invalidValue_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_leadingZerosInLongLengthArray1_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_leadingZerosInLongLengthArray2_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_leadingZerosInLongLengthList1_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_listLongerThan32Elements_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_listOfLists2_succeeds",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_listOfLists_succeeds",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_longList1_succeeds",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_longList2_succeeds",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_longListLessThan56Bytes_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_longStringLength_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_longStringLessThan56Bytes_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_multiList_succeeds",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_nonOptimalLongLengthArray1_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_nonOptimalLongLengthArray2_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_notEnoughContentForList1_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_notEnoughContentForList2_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_notEnoughContentForString1_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_notEnoughContentForString2_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_notLongEnough_reverts",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "test_readList_shortListMax1_succeeds",
    data: BytesLike
  ): Result;

  events: {
    "TransactionDeposited(address,address,uint256,bytes)": EventFragment;
    "log(string)": EventFragment;
    "log_address(address)": EventFragment;
    "log_array(uint256[])": EventFragment;
    "log_array(int256[])": EventFragment;
    "log_array(address[])": EventFragment;
    "log_bytes(bytes)": EventFragment;
    "log_bytes32(bytes32)": EventFragment;
    "log_int(int256)": EventFragment;
    "log_named_address(string,address)": EventFragment;
    "log_named_array(string,uint256[])": EventFragment;
    "log_named_array(string,int256[])": EventFragment;
    "log_named_array(string,address[])": EventFragment;
    "log_named_bytes(string,bytes)": EventFragment;
    "log_named_bytes32(string,bytes32)": EventFragment;
    "log_named_decimal_int(string,int256,uint256)": EventFragment;
    "log_named_decimal_uint(string,uint256,uint256)": EventFragment;
    "log_named_int(string,int256)": EventFragment;
    "log_named_string(string,string)": EventFragment;
    "log_named_uint(string,uint256)": EventFragment;
    "log_string(string)": EventFragment;
    "log_uint(uint256)": EventFragment;
    "logs(bytes)": EventFragment;
  };

  getEvent(nameOrSignatureOrTopic: "TransactionDeposited"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "log"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "log_address"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "log_array(uint256[])"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "log_array(int256[])"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "log_array(address[])"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "log_bytes"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "log_bytes32"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "log_int"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "log_named_address"): EventFragment;
  getEvent(
    nameOrSignatureOrTopic: "log_named_array(string,uint256[])"
  ): EventFragment;
  getEvent(
    nameOrSignatureOrTopic: "log_named_array(string,int256[])"
  ): EventFragment;
  getEvent(
    nameOrSignatureOrTopic: "log_named_array(string,address[])"
  ): EventFragment;
  getEvent(nameOrSignatureOrTopic: "log_named_bytes"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "log_named_bytes32"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "log_named_decimal_int"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "log_named_decimal_uint"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "log_named_int"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "log_named_string"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "log_named_uint"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "log_string"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "log_uint"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "logs"): EventFragment;
}

export interface TransactionDepositedEventObject {
  from: string;
  to: string;
  version: BigNumber;
  opaqueData: string;
}
export type TransactionDepositedEvent = TypedEvent<
  [string, string, BigNumber, string],
  TransactionDepositedEventObject
>;

export type TransactionDepositedEventFilter =
  TypedEventFilter<TransactionDepositedEvent>;

export interface logEventObject {
  arg0: string;
}
export type logEvent = TypedEvent<[string], logEventObject>;

export type logEventFilter = TypedEventFilter<logEvent>;

export interface log_addressEventObject {
  arg0: string;
}
export type log_addressEvent = TypedEvent<[string], log_addressEventObject>;

export type log_addressEventFilter = TypedEventFilter<log_addressEvent>;

export interface log_array_uint256_array_EventObject {
  val: BigNumber[];
}
export type log_array_uint256_array_Event = TypedEvent<
  [BigNumber[]],
  log_array_uint256_array_EventObject
>;

export type log_array_uint256_array_EventFilter =
  TypedEventFilter<log_array_uint256_array_Event>;

export interface log_array_int256_array_EventObject {
  val: BigNumber[];
}
export type log_array_int256_array_Event = TypedEvent<
  [BigNumber[]],
  log_array_int256_array_EventObject
>;

export type log_array_int256_array_EventFilter =
  TypedEventFilter<log_array_int256_array_Event>;

export interface log_array_address_array_EventObject {
  val: string[];
}
export type log_array_address_array_Event = TypedEvent<
  [string[]],
  log_array_address_array_EventObject
>;

export type log_array_address_array_EventFilter =
  TypedEventFilter<log_array_address_array_Event>;

export interface log_bytesEventObject {
  arg0: string;
}
export type log_bytesEvent = TypedEvent<[string], log_bytesEventObject>;

export type log_bytesEventFilter = TypedEventFilter<log_bytesEvent>;

export interface log_bytes32EventObject {
  arg0: string;
}
export type log_bytes32Event = TypedEvent<[string], log_bytes32EventObject>;

export type log_bytes32EventFilter = TypedEventFilter<log_bytes32Event>;

export interface log_intEventObject {
  arg0: BigNumber;
}
export type log_intEvent = TypedEvent<[BigNumber], log_intEventObject>;

export type log_intEventFilter = TypedEventFilter<log_intEvent>;

export interface log_named_addressEventObject {
  key: string;
  val: string;
}
export type log_named_addressEvent = TypedEvent<
  [string, string],
  log_named_addressEventObject
>;

export type log_named_addressEventFilter =
  TypedEventFilter<log_named_addressEvent>;

export interface log_named_array_string_uint256_array_EventObject {
  key: string;
  val: BigNumber[];
}
export type log_named_array_string_uint256_array_Event = TypedEvent<
  [string, BigNumber[]],
  log_named_array_string_uint256_array_EventObject
>;

export type log_named_array_string_uint256_array_EventFilter =
  TypedEventFilter<log_named_array_string_uint256_array_Event>;

export interface log_named_array_string_int256_array_EventObject {
  key: string;
  val: BigNumber[];
}
export type log_named_array_string_int256_array_Event = TypedEvent<
  [string, BigNumber[]],
  log_named_array_string_int256_array_EventObject
>;

export type log_named_array_string_int256_array_EventFilter =
  TypedEventFilter<log_named_array_string_int256_array_Event>;

export interface log_named_array_string_address_array_EventObject {
  key: string;
  val: string[];
}
export type log_named_array_string_address_array_Event = TypedEvent<
  [string, string[]],
  log_named_array_string_address_array_EventObject
>;

export type log_named_array_string_address_array_EventFilter =
  TypedEventFilter<log_named_array_string_address_array_Event>;

export interface log_named_bytesEventObject {
  key: string;
  val: string;
}
export type log_named_bytesEvent = TypedEvent<
  [string, string],
  log_named_bytesEventObject
>;

export type log_named_bytesEventFilter = TypedEventFilter<log_named_bytesEvent>;

export interface log_named_bytes32EventObject {
  key: string;
  val: string;
}
export type log_named_bytes32Event = TypedEvent<
  [string, string],
  log_named_bytes32EventObject
>;

export type log_named_bytes32EventFilter =
  TypedEventFilter<log_named_bytes32Event>;

export interface log_named_decimal_intEventObject {
  key: string;
  val: BigNumber;
  decimals: BigNumber;
}
export type log_named_decimal_intEvent = TypedEvent<
  [string, BigNumber, BigNumber],
  log_named_decimal_intEventObject
>;

export type log_named_decimal_intEventFilter =
  TypedEventFilter<log_named_decimal_intEvent>;

export interface log_named_decimal_uintEventObject {
  key: string;
  val: BigNumber;
  decimals: BigNumber;
}
export type log_named_decimal_uintEvent = TypedEvent<
  [string, BigNumber, BigNumber],
  log_named_decimal_uintEventObject
>;

export type log_named_decimal_uintEventFilter =
  TypedEventFilter<log_named_decimal_uintEvent>;

export interface log_named_intEventObject {
  key: string;
  val: BigNumber;
}
export type log_named_intEvent = TypedEvent<
  [string, BigNumber],
  log_named_intEventObject
>;

export type log_named_intEventFilter = TypedEventFilter<log_named_intEvent>;

export interface log_named_stringEventObject {
  key: string;
  val: string;
}
export type log_named_stringEvent = TypedEvent<
  [string, string],
  log_named_stringEventObject
>;

export type log_named_stringEventFilter =
  TypedEventFilter<log_named_stringEvent>;

export interface log_named_uintEventObject {
  key: string;
  val: BigNumber;
}
export type log_named_uintEvent = TypedEvent<
  [string, BigNumber],
  log_named_uintEventObject
>;

export type log_named_uintEventFilter = TypedEventFilter<log_named_uintEvent>;

export interface log_stringEventObject {
  arg0: string;
}
export type log_stringEvent = TypedEvent<[string], log_stringEventObject>;

export type log_stringEventFilter = TypedEventFilter<log_stringEvent>;

export interface log_uintEventObject {
  arg0: BigNumber;
}
export type log_uintEvent = TypedEvent<[BigNumber], log_uintEventObject>;

export type log_uintEventFilter = TypedEventFilter<log_uintEvent>;

export interface logsEventObject {
  arg0: string;
}
export type logsEvent = TypedEvent<[string], logsEventObject>;

export type logsEventFilter = TypedEventFilter<logsEvent>;

export interface RLPReader_readList_Test extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: RLPReader_readList_TestInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    IS_TEST(overrides?: CallOverrides): Promise<[boolean]>;

    excludeArtifacts(
      overrides?: CallOverrides
    ): Promise<[string[]] & { excludedArtifacts_: string[] }>;

    excludeContracts(
      overrides?: CallOverrides
    ): Promise<[string[]] & { excludedContracts_: string[] }>;

    excludeSenders(
      overrides?: CallOverrides
    ): Promise<[string[]] & { excludedSenders_: string[] }>;

    failed(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    setUp(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    targetArtifactSelectors(
      overrides?: CallOverrides
    ): Promise<
      [StdInvariant.FuzzSelectorStructOutput[]] & {
        targetedArtifactSelectors_: StdInvariant.FuzzSelectorStructOutput[];
      }
    >;

    targetArtifacts(
      overrides?: CallOverrides
    ): Promise<[string[]] & { targetedArtifacts_: string[] }>;

    targetContracts(
      overrides?: CallOverrides
    ): Promise<[string[]] & { targetedContracts_: string[] }>;

    targetSelectors(
      overrides?: CallOverrides
    ): Promise<
      [StdInvariant.FuzzSelectorStructOutput[]] & {
        targetedSelectors_: StdInvariant.FuzzSelectorStructOutput[];
      }
    >;

    targetSenders(
      overrides?: CallOverrides
    ): Promise<[string[]] & { targetedSenders_: string[] }>;

    test_readList_dictTest1_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_empty_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_incorrectLengthInArray_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_int32Overflow2_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_int32Overflow_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_invalidRemainder_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_invalidShortList_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_invalidValue_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_leadingZerosInLongLengthArray1_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_leadingZerosInLongLengthArray2_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_leadingZerosInLongLengthList1_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_listLongerThan32Elements_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_listOfLists2_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_listOfLists_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_longList1_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_longList2_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_longListLessThan56Bytes_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_longStringLength_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_longStringLessThan56Bytes_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_multiList_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_nonOptimalLongLengthArray1_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_nonOptimalLongLengthArray2_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_notEnoughContentForList1_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_notEnoughContentForList2_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_notEnoughContentForString1_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_notEnoughContentForString2_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_notLongEnough_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    test_readList_shortListMax1_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;
  };

  IS_TEST(overrides?: CallOverrides): Promise<boolean>;

  excludeArtifacts(overrides?: CallOverrides): Promise<string[]>;

  excludeContracts(overrides?: CallOverrides): Promise<string[]>;

  excludeSenders(overrides?: CallOverrides): Promise<string[]>;

  failed(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  setUp(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  targetArtifactSelectors(
    overrides?: CallOverrides
  ): Promise<StdInvariant.FuzzSelectorStructOutput[]>;

  targetArtifacts(overrides?: CallOverrides): Promise<string[]>;

  targetContracts(overrides?: CallOverrides): Promise<string[]>;

  targetSelectors(
    overrides?: CallOverrides
  ): Promise<StdInvariant.FuzzSelectorStructOutput[]>;

  targetSenders(overrides?: CallOverrides): Promise<string[]>;

  test_readList_dictTest1_succeeds(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_empty_succeeds(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_incorrectLengthInArray_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_int32Overflow2_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_int32Overflow_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_invalidRemainder_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_invalidShortList_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_invalidValue_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_leadingZerosInLongLengthArray1_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_leadingZerosInLongLengthArray2_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_leadingZerosInLongLengthList1_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_listLongerThan32Elements_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_listOfLists2_succeeds(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_listOfLists_succeeds(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_longList1_succeeds(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_longList2_succeeds(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_longListLessThan56Bytes_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_longStringLength_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_longStringLessThan56Bytes_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_multiList_succeeds(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_nonOptimalLongLengthArray1_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_nonOptimalLongLengthArray2_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_notEnoughContentForList1_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_notEnoughContentForList2_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_notEnoughContentForString1_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_notEnoughContentForString2_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_notLongEnough_reverts(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  test_readList_shortListMax1_succeeds(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    IS_TEST(overrides?: CallOverrides): Promise<boolean>;

    excludeArtifacts(overrides?: CallOverrides): Promise<string[]>;

    excludeContracts(overrides?: CallOverrides): Promise<string[]>;

    excludeSenders(overrides?: CallOverrides): Promise<string[]>;

    failed(overrides?: CallOverrides): Promise<boolean>;

    setUp(overrides?: CallOverrides): Promise<void>;

    targetArtifactSelectors(
      overrides?: CallOverrides
    ): Promise<StdInvariant.FuzzSelectorStructOutput[]>;

    targetArtifacts(overrides?: CallOverrides): Promise<string[]>;

    targetContracts(overrides?: CallOverrides): Promise<string[]>;

    targetSelectors(
      overrides?: CallOverrides
    ): Promise<StdInvariant.FuzzSelectorStructOutput[]>;

    targetSenders(overrides?: CallOverrides): Promise<string[]>;

    test_readList_dictTest1_succeeds(overrides?: CallOverrides): Promise<void>;

    test_readList_empty_succeeds(overrides?: CallOverrides): Promise<void>;

    test_readList_incorrectLengthInArray_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_int32Overflow2_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_int32Overflow_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_invalidRemainder_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_invalidShortList_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_invalidValue_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_leadingZerosInLongLengthArray1_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_leadingZerosInLongLengthArray2_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_leadingZerosInLongLengthList1_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_listLongerThan32Elements_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_listOfLists2_succeeds(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_listOfLists_succeeds(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_longList1_succeeds(overrides?: CallOverrides): Promise<void>;

    test_readList_longList2_succeeds(overrides?: CallOverrides): Promise<void>;

    test_readList_longListLessThan56Bytes_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_longStringLength_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_longStringLessThan56Bytes_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_multiList_succeeds(overrides?: CallOverrides): Promise<void>;

    test_readList_nonOptimalLongLengthArray1_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_nonOptimalLongLengthArray2_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_notEnoughContentForList1_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_notEnoughContentForList2_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_notEnoughContentForString1_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_notEnoughContentForString2_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_notLongEnough_reverts(
      overrides?: CallOverrides
    ): Promise<void>;

    test_readList_shortListMax1_succeeds(
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {
    "TransactionDeposited(address,address,uint256,bytes)"(
      from?: PromiseOrValue<string> | null,
      to?: PromiseOrValue<string> | null,
      version?: PromiseOrValue<BigNumberish> | null,
      opaqueData?: null
    ): TransactionDepositedEventFilter;
    TransactionDeposited(
      from?: PromiseOrValue<string> | null,
      to?: PromiseOrValue<string> | null,
      version?: PromiseOrValue<BigNumberish> | null,
      opaqueData?: null
    ): TransactionDepositedEventFilter;

    "log(string)"(arg0?: null): logEventFilter;
    log(arg0?: null): logEventFilter;

    "log_address(address)"(arg0?: null): log_addressEventFilter;
    log_address(arg0?: null): log_addressEventFilter;

    "log_array(uint256[])"(val?: null): log_array_uint256_array_EventFilter;
    "log_array(int256[])"(val?: null): log_array_int256_array_EventFilter;
    "log_array(address[])"(val?: null): log_array_address_array_EventFilter;

    "log_bytes(bytes)"(arg0?: null): log_bytesEventFilter;
    log_bytes(arg0?: null): log_bytesEventFilter;

    "log_bytes32(bytes32)"(arg0?: null): log_bytes32EventFilter;
    log_bytes32(arg0?: null): log_bytes32EventFilter;

    "log_int(int256)"(arg0?: null): log_intEventFilter;
    log_int(arg0?: null): log_intEventFilter;

    "log_named_address(string,address)"(
      key?: null,
      val?: null
    ): log_named_addressEventFilter;
    log_named_address(key?: null, val?: null): log_named_addressEventFilter;

    "log_named_array(string,uint256[])"(
      key?: null,
      val?: null
    ): log_named_array_string_uint256_array_EventFilter;
    "log_named_array(string,int256[])"(
      key?: null,
      val?: null
    ): log_named_array_string_int256_array_EventFilter;
    "log_named_array(string,address[])"(
      key?: null,
      val?: null
    ): log_named_array_string_address_array_EventFilter;

    "log_named_bytes(string,bytes)"(
      key?: null,
      val?: null
    ): log_named_bytesEventFilter;
    log_named_bytes(key?: null, val?: null): log_named_bytesEventFilter;

    "log_named_bytes32(string,bytes32)"(
      key?: null,
      val?: null
    ): log_named_bytes32EventFilter;
    log_named_bytes32(key?: null, val?: null): log_named_bytes32EventFilter;

    "log_named_decimal_int(string,int256,uint256)"(
      key?: null,
      val?: null,
      decimals?: null
    ): log_named_decimal_intEventFilter;
    log_named_decimal_int(
      key?: null,
      val?: null,
      decimals?: null
    ): log_named_decimal_intEventFilter;

    "log_named_decimal_uint(string,uint256,uint256)"(
      key?: null,
      val?: null,
      decimals?: null
    ): log_named_decimal_uintEventFilter;
    log_named_decimal_uint(
      key?: null,
      val?: null,
      decimals?: null
    ): log_named_decimal_uintEventFilter;

    "log_named_int(string,int256)"(
      key?: null,
      val?: null
    ): log_named_intEventFilter;
    log_named_int(key?: null, val?: null): log_named_intEventFilter;

    "log_named_string(string,string)"(
      key?: null,
      val?: null
    ): log_named_stringEventFilter;
    log_named_string(key?: null, val?: null): log_named_stringEventFilter;

    "log_named_uint(string,uint256)"(
      key?: null,
      val?: null
    ): log_named_uintEventFilter;
    log_named_uint(key?: null, val?: null): log_named_uintEventFilter;

    "log_string(string)"(arg0?: null): log_stringEventFilter;
    log_string(arg0?: null): log_stringEventFilter;

    "log_uint(uint256)"(arg0?: null): log_uintEventFilter;
    log_uint(arg0?: null): log_uintEventFilter;

    "logs(bytes)"(arg0?: null): logsEventFilter;
    logs(arg0?: null): logsEventFilter;
  };

  estimateGas: {
    IS_TEST(overrides?: CallOverrides): Promise<BigNumber>;

    excludeArtifacts(overrides?: CallOverrides): Promise<BigNumber>;

    excludeContracts(overrides?: CallOverrides): Promise<BigNumber>;

    excludeSenders(overrides?: CallOverrides): Promise<BigNumber>;

    failed(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    setUp(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    targetArtifactSelectors(overrides?: CallOverrides): Promise<BigNumber>;

    targetArtifacts(overrides?: CallOverrides): Promise<BigNumber>;

    targetContracts(overrides?: CallOverrides): Promise<BigNumber>;

    targetSelectors(overrides?: CallOverrides): Promise<BigNumber>;

    targetSenders(overrides?: CallOverrides): Promise<BigNumber>;

    test_readList_dictTest1_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_empty_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_incorrectLengthInArray_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_int32Overflow2_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_int32Overflow_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_invalidRemainder_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_invalidShortList_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_invalidValue_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_leadingZerosInLongLengthArray1_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_leadingZerosInLongLengthArray2_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_leadingZerosInLongLengthList1_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_listLongerThan32Elements_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_listOfLists2_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_listOfLists_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_longList1_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_longList2_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_longListLessThan56Bytes_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_longStringLength_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_longStringLessThan56Bytes_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_multiList_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_nonOptimalLongLengthArray1_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_nonOptimalLongLengthArray2_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_notEnoughContentForList1_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_notEnoughContentForList2_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_notEnoughContentForString1_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_notEnoughContentForString2_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_notLongEnough_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    test_readList_shortListMax1_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    IS_TEST(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    excludeArtifacts(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    excludeContracts(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    excludeSenders(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    failed(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    setUp(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    targetArtifactSelectors(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    targetArtifacts(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    targetContracts(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    targetSelectors(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    targetSenders(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    test_readList_dictTest1_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_empty_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_incorrectLengthInArray_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_int32Overflow2_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_int32Overflow_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_invalidRemainder_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_invalidShortList_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_invalidValue_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_leadingZerosInLongLengthArray1_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_leadingZerosInLongLengthArray2_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_leadingZerosInLongLengthList1_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_listLongerThan32Elements_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_listOfLists2_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_listOfLists_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_longList1_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_longList2_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_longListLessThan56Bytes_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_longStringLength_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_longStringLessThan56Bytes_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_multiList_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_nonOptimalLongLengthArray1_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_nonOptimalLongLengthArray2_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_notEnoughContentForList1_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_notEnoughContentForList2_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_notEnoughContentForString1_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_notEnoughContentForString2_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_notLongEnough_reverts(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    test_readList_shortListMax1_succeeds(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;
  };
}
