// @eth-optimism/contracts-bedrock/dist/types/contracts/echidna/FuzzAddressAliasing.sol/EchidnaFuzzAddressAliasing.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../../common";

export interface EchidnaFuzzAddressAliasingInterface extends utils.Interface {
  functions: {
    "echidna_round_trip_aliasing()": FunctionFragment;
    "testRoundTrip(address)": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic: "echidna_round_trip_aliasing" | "testRoundTrip"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "echidna_round_trip_aliasing",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "testRoundTrip",
    values: [PromiseOrValue<string>]
  ): string;

  decodeFunctionResult(
    functionFragment: "echidna_round_trip_aliasing",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "testRoundTrip",
    data: BytesLike
  ): Result;

  events: {};
}

export interface EchidnaFuzzAddressAliasing extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: EchidnaFuzzAddressAliasingInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    echidna_round_trip_aliasing(overrides?: CallOverrides): Promise<[boolean]>;

    testRoundTrip(
      addr: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;
  };

  echidna_round_trip_aliasing(overrides?: CallOverrides): Promise<boolean>;

  testRoundTrip(
    addr: PromiseOrValue<string>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    echidna_round_trip_aliasing(overrides?: CallOverrides): Promise<boolean>;

    testRoundTrip(
      addr: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {};

  estimateGas: {
    echidna_round_trip_aliasing(overrides?: CallOverrides): Promise<BigNumber>;

    testRoundTrip(
      addr: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    echidna_round_trip_aliasing(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    testRoundTrip(
      addr: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;
  };
}
