// @eth-optimism/contracts-bedrock/dist/types/contracts/test/invariants/SafeCall.t.sol/SafeCaller_Actor.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../../../common";

export interface SafeCaller_ActorInterface extends utils.Interface {
  functions: {
    "numCalls()": FunctionFragment;
    "performSafeCallMinGas(uint64,uint64,address,uint8)": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic: "numCalls" | "performSafeCallMinGas"
  ): FunctionFragment;

  encodeFunctionData(functionFragment: "numCalls", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "performSafeCallMinGas",
    values: [
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<string>,
      PromiseOrValue<BigNumberish>
    ]
  ): string;

  decodeFunctionResult(functionFragment: "numCalls", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "performSafeCallMinGas",
    data: BytesLike
  ): Result;

  events: {};
}

export interface SafeCaller_Actor extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: SafeCaller_ActorInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    numCalls(overrides?: CallOverrides): Promise<[BigNumber]>;

    performSafeCallMinGas(
      gas: PromiseOrValue<BigNumberish>,
      minGas: PromiseOrValue<BigNumberish>,
      to: PromiseOrValue<string>,
      value: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;
  };

  numCalls(overrides?: CallOverrides): Promise<BigNumber>;

  performSafeCallMinGas(
    gas: PromiseOrValue<BigNumberish>,
    minGas: PromiseOrValue<BigNumberish>,
    to: PromiseOrValue<string>,
    value: PromiseOrValue<BigNumberish>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    numCalls(overrides?: CallOverrides): Promise<BigNumber>;

    performSafeCallMinGas(
      gas: PromiseOrValue<BigNumberish>,
      minGas: PromiseOrValue<BigNumberish>,
      to: PromiseOrValue<string>,
      value: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {};

  estimateGas: {
    numCalls(overrides?: CallOverrides): Promise<BigNumber>;

    performSafeCallMinGas(
      gas: PromiseOrValue<BigNumberish>,
      minGas: PromiseOrValue<BigNumberish>,
      to: PromiseOrValue<string>,
      value: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    numCalls(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    performSafeCallMinGas(
      gas: PromiseOrValue<BigNumberish>,
      minGas: PromiseOrValue<BigNumberish>,
      to: PromiseOrValue<string>,
      value: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;
  };
}
