// @eth-optimism/contracts-bedrock/dist/types/contracts/deployment/PortalSender.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../common";

export interface PortalSenderInterface extends utils.Interface {
  functions: {
    "PORTAL()": FunctionFragment;
    "donate()": FunctionFragment;
  };

  getFunction(nameOrSignatureOrTopic: "PORTAL" | "donate"): FunctionFragment;

  encodeFunctionData(functionFragment: "PORTAL", values?: undefined): string;
  encodeFunctionData(functionFragment: "donate", values?: undefined): string;

  decodeFunctionResult(functionFragment: "PORTAL", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "donate", data: BytesLike): Result;

  events: {};
}

export interface PortalSender extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: PortalSenderInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    PORTAL(overrides?: CallOverrides): Promise<[string]>;

    donate(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;
  };

  PORTAL(overrides?: CallOverrides): Promise<string>;

  donate(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    PORTAL(overrides?: CallOverrides): Promise<string>;

    donate(overrides?: CallOverrides): Promise<void>;
  };

  filters: {};

  estimateGas: {
    PORTAL(overrides?: CallOverrides): Promise<BigNumber>;

    donate(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    PORTAL(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    donate(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;
  };
}
