// @eth-optimism/contracts-bedrock/dist/types/contracts/test/invariants/Burn.Eth.t.sol/Burn_EthBurner.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../../../common";

export interface Burn_EthBurnerInterface extends utils.Interface {
  functions: {
    "burnEth(uint256)": FunctionFragment;
    "failedEthBurn()": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic: "burnEth" | "failedEthBurn"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "burnEth",
    values: [PromiseOrValue<BigNumberish>]
  ): string;
  encodeFunctionData(
    functionFragment: "failedEthBurn",
    values?: undefined
  ): string;

  decodeFunctionResult(functionFragment: "burnEth", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "failedEthBurn",
    data: BytesLike
  ): Result;

  events: {};
}

export interface Burn_EthBurner extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: Burn_EthBurnerInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    burnEth(
      _value: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    failedEthBurn(overrides?: CallOverrides): Promise<[boolean]>;
  };

  burnEth(
    _value: PromiseOrValue<BigNumberish>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  failedEthBurn(overrides?: CallOverrides): Promise<boolean>;

  callStatic: {
    burnEth(
      _value: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<void>;

    failedEthBurn(overrides?: CallOverrides): Promise<boolean>;
  };

  filters: {};

  estimateGas: {
    burnEth(
      _value: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    failedEthBurn(overrides?: CallOverrides): Promise<BigNumber>;
  };

  populateTransaction: {
    burnEth(
      _value: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    failedEthBurn(overrides?: CallOverrides): Promise<PopulatedTransaction>;
  };
}
