// @eth-optimism/contracts-bedrock/dist/types/factories/contracts/dispute/IDisputeGame__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  IDisputeGame,
  IDisputeGameInterface,
} from "../../../contracts/dispute/IDisputeGame";

const _abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "enum GameStatus",
        name: "status",
        type: "uint8",
      },
    ],
    name: "Resolved",
    type: "event",
  },
  {
    inputs: [],
    name: "bondManager",
    outputs: [
      {
        internalType: "contract IBondManager",
        name: "_bondManager",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "createdAt",
    outputs: [
      {
        internalType: "Timestamp",
        name: "_createdAt",
        type: "uint64",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "extraData",
    outputs: [
      {
        internalType: "bytes",
        name: "_extraData",
        type: "bytes",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "gameType",
    outputs: [
      {
        internalType: "enum GameType",
        name: "_gameType",
        type: "uint8",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "initialize",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [],
    name: "resolve",
    outputs: [
      {
        internalType: "enum GameStatus",
        name: "_status",
        type: "uint8",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [],
    name: "rootClaim",
    outputs: [
      {
        internalType: "Claim",
        name: "_rootClaim",
        type: "bytes32",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "status",
    outputs: [
      {
        internalType: "enum GameStatus",
        name: "_status",
        type: "uint8",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "version",
    outputs: [
      {
        internalType: "string",
        name: "_version",
        type: "string",
      },
    ],
    stateMutability: "pure",
    type: "function",
  },
];

export class IDisputeGame__factory {
  static readonly abi = _abi;
  static createInterface(): IDisputeGameInterface {
    return new utils.Interface(_abi) as IDisputeGameInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IDisputeGame {
    return new Contract(address, _abi, signerOrProvider) as IDisputeGame;
  }
}
