// @eth-optimism/contracts-bedrock/dist/types/contracts/dispute/IDisputeGameFactory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type {
  FunctionFragment,
  Result,
  EventFragment,
} from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../common";

export interface IDisputeGameFactoryInterface extends utils.Interface {
  functions: {
    "create(uint8,bytes32,bytes)": FunctionFragment;
    "gameImpls(uint8)": FunctionFragment;
    "games(uint8,bytes32,bytes)": FunctionFragment;
    "setImplementation(uint8,address)": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "create"
      | "gameImpls"
      | "games"
      | "setImplementation"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "create",
    values: [
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<BytesLike>,
      PromiseOrValue<BytesLike>
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "gameImpls",
    values: [PromiseOrValue<BigNumberish>]
  ): string;
  encodeFunctionData(
    functionFragment: "games",
    values: [
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<BytesLike>,
      PromiseOrValue<BytesLike>
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "setImplementation",
    values: [PromiseOrValue<BigNumberish>, PromiseOrValue<string>]
  ): string;

  decodeFunctionResult(functionFragment: "create", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "gameImpls", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "games", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "setImplementation",
    data: BytesLike
  ): Result;

  events: {
    "DisputeGameCreated(address,uint8,bytes32)": EventFragment;
    "ImplementationSet(address,uint8)": EventFragment;
  };

  getEvent(nameOrSignatureOrTopic: "DisputeGameCreated"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "ImplementationSet"): EventFragment;
}

export interface DisputeGameCreatedEventObject {
  disputeProxy: string;
  gameType: number;
  rootClaim: string;
}
export type DisputeGameCreatedEvent = TypedEvent<
  [string, number, string],
  DisputeGameCreatedEventObject
>;

export type DisputeGameCreatedEventFilter =
  TypedEventFilter<DisputeGameCreatedEvent>;

export interface ImplementationSetEventObject {
  impl: string;
  gameType: number;
}
export type ImplementationSetEvent = TypedEvent<
  [string, number],
  ImplementationSetEventObject
>;

export type ImplementationSetEventFilter =
  TypedEventFilter<ImplementationSetEvent>;

export interface IDisputeGameFactory extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: IDisputeGameFactoryInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    create(
      gameType: PromiseOrValue<BigNumberish>,
      rootClaim: PromiseOrValue<BytesLike>,
      extraData: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    gameImpls(
      gameType: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<[string] & { _impl: string }>;

    games(
      gameType: PromiseOrValue<BigNumberish>,
      rootClaim: PromiseOrValue<BytesLike>,
      extraData: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<[string] & { _proxy: string }>;

    setImplementation(
      gameType: PromiseOrValue<BigNumberish>,
      impl: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;
  };

  create(
    gameType: PromiseOrValue<BigNumberish>,
    rootClaim: PromiseOrValue<BytesLike>,
    extraData: PromiseOrValue<BytesLike>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  gameImpls(
    gameType: PromiseOrValue<BigNumberish>,
    overrides?: CallOverrides
  ): Promise<string>;

  games(
    gameType: PromiseOrValue<BigNumberish>,
    rootClaim: PromiseOrValue<BytesLike>,
    extraData: PromiseOrValue<BytesLike>,
    overrides?: CallOverrides
  ): Promise<string>;

  setImplementation(
    gameType: PromiseOrValue<BigNumberish>,
    impl: PromiseOrValue<string>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    create(
      gameType: PromiseOrValue<BigNumberish>,
      rootClaim: PromiseOrValue<BytesLike>,
      extraData: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<string>;

    gameImpls(
      gameType: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<string>;

    games(
      gameType: PromiseOrValue<BigNumberish>,
      rootClaim: PromiseOrValue<BytesLike>,
      extraData: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<string>;

    setImplementation(
      gameType: PromiseOrValue<BigNumberish>,
      impl: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {
    "DisputeGameCreated(address,uint8,bytes32)"(
      disputeProxy?: PromiseOrValue<string> | null,
      gameType?: PromiseOrValue<BigNumberish> | null,
      rootClaim?: PromiseOrValue<BytesLike> | null
    ): DisputeGameCreatedEventFilter;
    DisputeGameCreated(
      disputeProxy?: PromiseOrValue<string> | null,
      gameType?: PromiseOrValue<BigNumberish> | null,
      rootClaim?: PromiseOrValue<BytesLike> | null
    ): DisputeGameCreatedEventFilter;

    "ImplementationSet(address,uint8)"(
      impl?: PromiseOrValue<string> | null,
      gameType?: PromiseOrValue<BigNumberish> | null
    ): ImplementationSetEventFilter;
    ImplementationSet(
      impl?: PromiseOrValue<string> | null,
      gameType?: PromiseOrValue<BigNumberish> | null
    ): ImplementationSetEventFilter;
  };

  estimateGas: {
    create(
      gameType: PromiseOrValue<BigNumberish>,
      rootClaim: PromiseOrValue<BytesLike>,
      extraData: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    gameImpls(
      gameType: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    games(
      gameType: PromiseOrValue<BigNumberish>,
      rootClaim: PromiseOrValue<BytesLike>,
      extraData: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    setImplementation(
      gameType: PromiseOrValue<BigNumberish>,
      impl: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    create(
      gameType: PromiseOrValue<BigNumberish>,
      rootClaim: PromiseOrValue<BytesLike>,
      extraData: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    gameImpls(
      gameType: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    games(
      gameType: PromiseOrValue<BigNumberish>,
      rootClaim: PromiseOrValue<BytesLike>,
      extraData: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    setImplementation(
      gameType: PromiseOrValue<BigNumberish>,
      impl: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;
  };
}
