// @eth-optimism/contracts-bedrock/dist/types/contracts/dispute/IFaultDisputeGame.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type {
  FunctionFragment,
  Result,
  EventFragment,
} from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../common";

export interface IFaultDisputeGameInterface extends utils.Interface {
  functions: {
    "attack(bytes32,bytes32)": FunctionFragment;
    "bondManager()": FunctionFragment;
    "bonds(bytes32)": FunctionFragment;
    "claims(bytes32)": FunctionFragment;
    "clocks(bytes32)": FunctionFragment;
    "countered(bytes32)": FunctionFragment;
    "createdAt()": FunctionFragment;
    "defend(bytes32,bytes32)": FunctionFragment;
    "extraData()": FunctionFragment;
    "gameStart()": FunctionFragment;
    "gameType()": FunctionFragment;
    "initialize()": FunctionFragment;
    "parents(bytes32)": FunctionFragment;
    "positions(bytes32)": FunctionFragment;
    "rc(bytes32)": FunctionFragment;
    "resolve()": FunctionFragment;
    "rootClaim()": FunctionFragment;
    "status()": FunctionFragment;
    "step(bytes32)": FunctionFragment;
    "version()": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "attack"
      | "bondManager"
      | "bonds"
      | "claims"
      | "clocks"
      | "countered"
      | "createdAt"
      | "defend"
      | "extraData"
      | "gameStart"
      | "gameType"
      | "initialize"
      | "parents"
      | "positions"
      | "rc"
      | "resolve"
      | "rootClaim"
      | "status"
      | "step"
      | "version"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "attack",
    values: [PromiseOrValue<BytesLike>, PromiseOrValue<BytesLike>]
  ): string;
  encodeFunctionData(
    functionFragment: "bondManager",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "bonds",
    values: [PromiseOrValue<BytesLike>]
  ): string;
  encodeFunctionData(
    functionFragment: "claims",
    values: [PromiseOrValue<BytesLike>]
  ): string;
  encodeFunctionData(
    functionFragment: "clocks",
    values: [PromiseOrValue<BytesLike>]
  ): string;
  encodeFunctionData(
    functionFragment: "countered",
    values: [PromiseOrValue<BytesLike>]
  ): string;
  encodeFunctionData(functionFragment: "createdAt", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "defend",
    values: [PromiseOrValue<BytesLike>, PromiseOrValue<BytesLike>]
  ): string;
  encodeFunctionData(functionFragment: "extraData", values?: undefined): string;
  encodeFunctionData(functionFragment: "gameStart", values?: undefined): string;
  encodeFunctionData(functionFragment: "gameType", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "initialize",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "parents",
    values: [PromiseOrValue<BytesLike>]
  ): string;
  encodeFunctionData(
    functionFragment: "positions",
    values: [PromiseOrValue<BytesLike>]
  ): string;
  encodeFunctionData(
    functionFragment: "rc",
    values: [PromiseOrValue<BytesLike>]
  ): string;
  encodeFunctionData(functionFragment: "resolve", values?: undefined): string;
  encodeFunctionData(functionFragment: "rootClaim", values?: undefined): string;
  encodeFunctionData(functionFragment: "status", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "step",
    values: [PromiseOrValue<BytesLike>]
  ): string;
  encodeFunctionData(functionFragment: "version", values?: undefined): string;

  decodeFunctionResult(functionFragment: "attack", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "bondManager",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "bonds", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "claims", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "clocks", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "countered", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "createdAt", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "defend", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "extraData", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "gameStart", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "gameType", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "initialize", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "parents", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "positions", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "rc", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "resolve", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "rootClaim", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "status", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "step", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "version", data: BytesLike): Result;

  events: {
    "Attack(bytes32,bytes32,address)": EventFragment;
    "Defend(bytes32,bytes32,address)": EventFragment;
    "Resolved(uint8)": EventFragment;
  };

  getEvent(nameOrSignatureOrTopic: "Attack"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "Defend"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "Resolved"): EventFragment;
}

export interface AttackEventObject {
  claimHash: string;
  pivot: string;
  claimant: string;
}
export type AttackEvent = TypedEvent<
  [string, string, string],
  AttackEventObject
>;

export type AttackEventFilter = TypedEventFilter<AttackEvent>;

export interface DefendEventObject {
  claimHash: string;
  pivot: string;
  claimant: string;
}
export type DefendEvent = TypedEvent<
  [string, string, string],
  DefendEventObject
>;

export type DefendEventFilter = TypedEventFilter<DefendEvent>;

export interface ResolvedEventObject {
  status: number;
}
export type ResolvedEvent = TypedEvent<[number], ResolvedEventObject>;

export type ResolvedEventFilter = TypedEventFilter<ResolvedEvent>;

export interface IFaultDisputeGame extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: IFaultDisputeGameInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    attack(
      disagreement: PromiseOrValue<BytesLike>,
      pivot: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    bondManager(
      overrides?: CallOverrides
    ): Promise<[string] & { _bondManager: string }>;

    bonds(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<[BigNumber] & { bond: BigNumber }>;

    claims(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<[string] & { claim: string }>;

    clocks(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<[BigNumber] & { clock: BigNumber }>;

    countered(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<[boolean] & { _countered: boolean }>;

    createdAt(
      overrides?: CallOverrides
    ): Promise<[BigNumber] & { _createdAt: BigNumber }>;

    defend(
      agreement: PromiseOrValue<BytesLike>,
      pivot: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    extraData(
      overrides?: CallOverrides
    ): Promise<[string] & { _extraData: string }>;

    gameStart(overrides?: CallOverrides): Promise<[BigNumber]>;

    gameType(
      overrides?: CallOverrides
    ): Promise<[number] & { _gameType: number }>;

    initialize(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    parents(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<[string] & { parent: string }>;

    positions(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<[BigNumber] & { position: BigNumber }>;

    rc(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<[BigNumber] & { _rc: BigNumber }>;

    resolve(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    rootClaim(
      overrides?: CallOverrides
    ): Promise<[string] & { _rootClaim: string }>;

    status(overrides?: CallOverrides): Promise<[number] & { _status: number }>;

    step(
      disagreement: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    version(
      overrides?: CallOverrides
    ): Promise<[string] & { _version: string }>;
  };

  attack(
    disagreement: PromiseOrValue<BytesLike>,
    pivot: PromiseOrValue<BytesLike>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  bondManager(overrides?: CallOverrides): Promise<string>;

  bonds(
    claimHash: PromiseOrValue<BytesLike>,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  claims(
    claimHash: PromiseOrValue<BytesLike>,
    overrides?: CallOverrides
  ): Promise<string>;

  clocks(
    claimHash: PromiseOrValue<BytesLike>,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  countered(
    claimHash: PromiseOrValue<BytesLike>,
    overrides?: CallOverrides
  ): Promise<boolean>;

  createdAt(overrides?: CallOverrides): Promise<BigNumber>;

  defend(
    agreement: PromiseOrValue<BytesLike>,
    pivot: PromiseOrValue<BytesLike>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  extraData(overrides?: CallOverrides): Promise<string>;

  gameStart(overrides?: CallOverrides): Promise<BigNumber>;

  gameType(overrides?: CallOverrides): Promise<number>;

  initialize(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  parents(
    claimHash: PromiseOrValue<BytesLike>,
    overrides?: CallOverrides
  ): Promise<string>;

  positions(
    claimHash: PromiseOrValue<BytesLike>,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  rc(
    claimHash: PromiseOrValue<BytesLike>,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  resolve(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  rootClaim(overrides?: CallOverrides): Promise<string>;

  status(overrides?: CallOverrides): Promise<number>;

  step(
    disagreement: PromiseOrValue<BytesLike>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  version(overrides?: CallOverrides): Promise<string>;

  callStatic: {
    attack(
      disagreement: PromiseOrValue<BytesLike>,
      pivot: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<void>;

    bondManager(overrides?: CallOverrides): Promise<string>;

    bonds(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    claims(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<string>;

    clocks(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    countered(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<boolean>;

    createdAt(overrides?: CallOverrides): Promise<BigNumber>;

    defend(
      agreement: PromiseOrValue<BytesLike>,
      pivot: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<void>;

    extraData(overrides?: CallOverrides): Promise<string>;

    gameStart(overrides?: CallOverrides): Promise<BigNumber>;

    gameType(overrides?: CallOverrides): Promise<number>;

    initialize(overrides?: CallOverrides): Promise<void>;

    parents(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<string>;

    positions(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    rc(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    resolve(overrides?: CallOverrides): Promise<number>;

    rootClaim(overrides?: CallOverrides): Promise<string>;

    status(overrides?: CallOverrides): Promise<number>;

    step(
      disagreement: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<void>;

    version(overrides?: CallOverrides): Promise<string>;
  };

  filters: {
    "Attack(bytes32,bytes32,address)"(
      claimHash?: PromiseOrValue<BytesLike> | null,
      pivot?: PromiseOrValue<BytesLike> | null,
      claimant?: PromiseOrValue<string> | null
    ): AttackEventFilter;
    Attack(
      claimHash?: PromiseOrValue<BytesLike> | null,
      pivot?: PromiseOrValue<BytesLike> | null,
      claimant?: PromiseOrValue<string> | null
    ): AttackEventFilter;

    "Defend(bytes32,bytes32,address)"(
      claimHash?: PromiseOrValue<BytesLike> | null,
      pivot?: PromiseOrValue<BytesLike> | null,
      claimant?: PromiseOrValue<string> | null
    ): DefendEventFilter;
    Defend(
      claimHash?: PromiseOrValue<BytesLike> | null,
      pivot?: PromiseOrValue<BytesLike> | null,
      claimant?: PromiseOrValue<string> | null
    ): DefendEventFilter;

    "Resolved(uint8)"(
      status?: PromiseOrValue<BigNumberish> | null
    ): ResolvedEventFilter;
    Resolved(status?: PromiseOrValue<BigNumberish> | null): ResolvedEventFilter;
  };

  estimateGas: {
    attack(
      disagreement: PromiseOrValue<BytesLike>,
      pivot: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    bondManager(overrides?: CallOverrides): Promise<BigNumber>;

    bonds(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    claims(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    clocks(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    countered(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    createdAt(overrides?: CallOverrides): Promise<BigNumber>;

    defend(
      agreement: PromiseOrValue<BytesLike>,
      pivot: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    extraData(overrides?: CallOverrides): Promise<BigNumber>;

    gameStart(overrides?: CallOverrides): Promise<BigNumber>;

    gameType(overrides?: CallOverrides): Promise<BigNumber>;

    initialize(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    parents(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    positions(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    rc(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    resolve(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    rootClaim(overrides?: CallOverrides): Promise<BigNumber>;

    status(overrides?: CallOverrides): Promise<BigNumber>;

    step(
      disagreement: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    version(overrides?: CallOverrides): Promise<BigNumber>;
  };

  populateTransaction: {
    attack(
      disagreement: PromiseOrValue<BytesLike>,
      pivot: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    bondManager(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    bonds(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    claims(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    clocks(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    countered(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    createdAt(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    defend(
      agreement: PromiseOrValue<BytesLike>,
      pivot: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    extraData(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    gameStart(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    gameType(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    initialize(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    parents(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    positions(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    rc(
      claimHash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    resolve(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    rootClaim(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    status(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    step(
      disagreement: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    version(overrides?: CallOverrides): Promise<PopulatedTransaction>;
  };
}
