// @eth-optimism/contracts-bedrock/dist/types/contracts/periphery/TransferOnion.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../common";

export declare namespace TransferOnion {
  export type LayerStruct = {
    recipient: PromiseOrValue<string>;
    amount: PromiseOrValue<BigNumberish>;
    shell: PromiseOrValue<BytesLike>;
  };

  export type LayerStructOutput = [string, BigNumber, string] & {
    recipient: string;
    amount: BigNumber;
    shell: string;
  };
}

export interface TransferOnionInterface extends utils.Interface {
  functions: {
    "SENDER()": FunctionFragment;
    "TOKEN()": FunctionFragment;
    "peel((address,uint256,bytes32)[])": FunctionFragment;
    "shell()": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic: "SENDER" | "TOKEN" | "peel" | "shell"
  ): FunctionFragment;

  encodeFunctionData(functionFragment: "SENDER", values?: undefined): string;
  encodeFunctionData(functionFragment: "TOKEN", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "peel",
    values: [TransferOnion.LayerStruct[]]
  ): string;
  encodeFunctionData(functionFragment: "shell", values?: undefined): string;

  decodeFunctionResult(functionFragment: "SENDER", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "TOKEN", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "peel", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "shell", data: BytesLike): Result;

  events: {};
}

export interface TransferOnion extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: TransferOnionInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    SENDER(overrides?: CallOverrides): Promise<[string]>;

    TOKEN(overrides?: CallOverrides): Promise<[string]>;

    peel(
      _layers: TransferOnion.LayerStruct[],
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    shell(overrides?: CallOverrides): Promise<[string]>;
  };

  SENDER(overrides?: CallOverrides): Promise<string>;

  TOKEN(overrides?: CallOverrides): Promise<string>;

  peel(
    _layers: TransferOnion.LayerStruct[],
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  shell(overrides?: CallOverrides): Promise<string>;

  callStatic: {
    SENDER(overrides?: CallOverrides): Promise<string>;

    TOKEN(overrides?: CallOverrides): Promise<string>;

    peel(
      _layers: TransferOnion.LayerStruct[],
      overrides?: CallOverrides
    ): Promise<void>;

    shell(overrides?: CallOverrides): Promise<string>;
  };

  filters: {};

  estimateGas: {
    SENDER(overrides?: CallOverrides): Promise<BigNumber>;

    TOKEN(overrides?: CallOverrides): Promise<BigNumber>;

    peel(
      _layers: TransferOnion.LayerStruct[],
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    shell(overrides?: CallOverrides): Promise<BigNumber>;
  };

  populateTransaction: {
    SENDER(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    TOKEN(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    peel(
      _layers: TransferOnion.LayerStruct[],
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    shell(overrides?: CallOverrides): Promise<PopulatedTransaction>;
  };
}
