// @eth-optimism/contracts-bedrock/dist/types/contracts/echidna/FuzzBurn.sol/EchidnaFuzzBurnGas.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../../common";

export interface EchidnaFuzzBurnGasInterface extends utils.Interface {
  functions: {
    "echidna_burn_gas()": FunctionFragment;
    "testGas(uint256)": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic: "echidna_burn_gas" | "testGas"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "echidna_burn_gas",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "testGas",
    values: [PromiseOrValue<BigNumberish>]
  ): string;

  decodeFunctionResult(
    functionFragment: "echidna_burn_gas",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "testGas", data: BytesLike): Result;

  events: {};
}

export interface EchidnaFuzzBurnGas extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: EchidnaFuzzBurnGasInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    echidna_burn_gas(overrides?: CallOverrides): Promise<[boolean]>;

    testGas(
      _value: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;
  };

  echidna_burn_gas(overrides?: CallOverrides): Promise<boolean>;

  testGas(
    _value: PromiseOrValue<BigNumberish>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    echidna_burn_gas(overrides?: CallOverrides): Promise<boolean>;

    testGas(
      _value: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {};

  estimateGas: {
    echidna_burn_gas(overrides?: CallOverrides): Promise<BigNumber>;

    testGas(
      _value: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    echidna_burn_gas(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    testGas(
      _value: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;
  };
}
