// @eth-optimism/contracts-bedrock/dist/types/contracts/echidna/FuzzEncoding.sol/EchidnaFuzzEncoding.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../../common";

export interface EchidnaFuzzEncodingInterface extends utils.Interface {
  functions: {
    "echidna_round_trip_encoding_AToB()": FunctionFragment;
    "echidna_round_trip_encoding_BToA()": FunctionFragment;
    "testRoundTripAToB(uint240,uint16)": FunctionFragment;
    "testRoundTripBToA(uint256)": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "echidna_round_trip_encoding_AToB"
      | "echidna_round_trip_encoding_BToA"
      | "testRoundTripAToB"
      | "testRoundTripBToA"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "echidna_round_trip_encoding_AToB",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "echidna_round_trip_encoding_BToA",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "testRoundTripAToB",
    values: [PromiseOrValue<BigNumberish>, PromiseOrValue<BigNumberish>]
  ): string;
  encodeFunctionData(
    functionFragment: "testRoundTripBToA",
    values: [PromiseOrValue<BigNumberish>]
  ): string;

  decodeFunctionResult(
    functionFragment: "echidna_round_trip_encoding_AToB",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "echidna_round_trip_encoding_BToA",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "testRoundTripAToB",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "testRoundTripBToA",
    data: BytesLike
  ): Result;

  events: {};
}

export interface EchidnaFuzzEncoding extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: EchidnaFuzzEncodingInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    echidna_round_trip_encoding_AToB(
      overrides?: CallOverrides
    ): Promise<[boolean]>;

    echidna_round_trip_encoding_BToA(
      overrides?: CallOverrides
    ): Promise<[boolean]>;

    testRoundTripAToB(
      _nonce: PromiseOrValue<BigNumberish>,
      _version: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    testRoundTripBToA(
      _versionedNonce: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;
  };

  echidna_round_trip_encoding_AToB(overrides?: CallOverrides): Promise<boolean>;

  echidna_round_trip_encoding_BToA(overrides?: CallOverrides): Promise<boolean>;

  testRoundTripAToB(
    _nonce: PromiseOrValue<BigNumberish>,
    _version: PromiseOrValue<BigNumberish>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  testRoundTripBToA(
    _versionedNonce: PromiseOrValue<BigNumberish>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    echidna_round_trip_encoding_AToB(
      overrides?: CallOverrides
    ): Promise<boolean>;

    echidna_round_trip_encoding_BToA(
      overrides?: CallOverrides
    ): Promise<boolean>;

    testRoundTripAToB(
      _nonce: PromiseOrValue<BigNumberish>,
      _version: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<void>;

    testRoundTripBToA(
      _versionedNonce: PromiseOrValue<BigNumberish>,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {};

  estimateGas: {
    echidna_round_trip_encoding_AToB(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    echidna_round_trip_encoding_BToA(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    testRoundTripAToB(
      _nonce: PromiseOrValue<BigNumberish>,
      _version: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    testRoundTripBToA(
      _versionedNonce: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    echidna_round_trip_encoding_AToB(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    echidna_round_trip_encoding_BToA(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    testRoundTripAToB(
      _nonce: PromiseOrValue<BigNumberish>,
      _version: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    testRoundTripBToA(
      _versionedNonce: PromiseOrValue<BigNumberish>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;
  };
}
